python/lsst/ts/tcpip/one_client_server.py

```diff
@@ -36,30 +41,34 @@ class OneClientServer:
 
     Parameters
     ----------
-    name : `str`
-        Name used for error messages. Typically "Commands" or "Telemetry".
     host : `str` or `None`
-        IP address for this server; typically `LOCALHOST` for IP4
-        or "::" for IP6. If `None` then bind to all network interfaces
-        (e.g. listen on an IPv4 socket and an IPv6 socket).
+        IP address for this server; typically `LOCALHOST_IPV4` for IPV4
+        or `LOCALHOST_IPV6` for IPV6. If `None` then bind to all network
+        interfaces (e.g. listen on an IPv4 socket and an IPv6 socket).
         None can cause trouble with port=0;
-        see port in Attributes for more information.
+        see ``port`` in the Attributes section for more information.
     port : `int`
         IP port for this server. If 0 then randomly pick an available port
         (or ports, if listening on multiple sockets).
         0 is strongly recommended for unit tests.
     log : `logging.Logger`
         Logger.
     connect_callback : callable or `None`, optional
-        Synchronous function to call when the connection state changes.
+        Asynchronous function to call when the connection state changes.
         If the client closes the connection it may take up to 0.1 seconds
         to notice (see Notes).
         It receives one argument: this `OneClientServer`.
```

Comment:
It is certainly possible to support both (and indeed OneClientServer does, for backwards compatibility, but that is deprecated). I lean towards requiring async because:
* It simplifies and regularizes the API
* Any sync function can trivially be made async by redefining it (though it requires a trivial wrapper for functions that you can't change, such as those from 3rd party libraries).
* I consider it a minor design flaw that salobj's ReadTopic supports both. When salobj allowed synchronous writes this made more sense, but now there are many fewer use cases for a synchronous callback. I have been considering deprecating support for synchronous callbacks in salobj and removing them entirely in the Kafka version. So now is a good time to complain about that, if you want me to keep the old behavior.

By the way: you should use `inspect.iscoroutinefunction` instead of `asyncio.iscoroutinefunction`. The asyncio version is limited and intended for internal use to handle a specific use case. The `inspect` version is more robust and intended for general use.  Another approach that salobj ReadTopic uses is to test the result of calling the callback to see if it is awaitable. I consider that slightly more robust (a user can't fool it by changing the callback function).

Regarding *args, **kwargs: I strongly prefer to follow Python's lead in this area and ask the user to "curry" any function that requires extra arguments by using `functools.partial`. One could go as far as to have no arguments at all, and I sometimes wish I had done that. But it is sometimes useful to provide the object whose state change triggered the callback (even though functools.partial could be used to handle that case, as well), and it seems a reasonable compromise to me. Certainly it's what I was already doing, so I am reluctant to change it.

Reply:
I think the `asyncio.iscoroutinefunction()` calls the `inspect.iscoroutinefunction()` already:
https://github.com/python/cpython/blob/3.11/Lib/asyncio/coroutines.py#L21-L24